packages/interface-stream-muxer/src/stream.ts

```diff
@@ -0,0 +1,321 @@
+  /**
+   * A unique identifier for this stream
+   */
+  id: string
+
+  /**
+   * The stream direction
+   */
+  direction: Direction
+
+  /**
+   * The maximum allowable data size, any data larger than this will be
+   * chunked and sent in multiple data messages
+   */
+  maxDataSize: number
+
+  /**
+   * User specific stream metadata
+   */
+  metadata?: Record<string, unknown>
+
+  /**
+   * Invoked when the stream ends
+   */
+  onEnd?: (err?: Error | undefined) => void
+}
+
+export abstract class AbstractStream implements Stream {
+  public id: string
+  public stat: StreamStat
+  public metadata: Record<string, unknown>
+  public source: AsyncGenerator<Uint8ArrayList, void, unknown>
+
+  private readonly abortController: AbortController
+  private readonly resetController: AbortController
+  private readonly closeController: AbortController
+  private sourceEnded: boolean
+  private sinkEnded: boolean
+  private sinkSunk: boolean
+  private endErr: Error | undefined
+  private readonly streamSource: Pushable<Uint8ArrayList>
+  private readonly onEnd?: (err?: Error | undefined) => void
+  private readonly maxDataSize: number
+
+  constructor (init: AbstractStreamInit) {
+    this.abortController = new AbortController()
+    this.resetController = new AbortController()
+    this.closeController = new AbortController()
+    this.sourceEnded = false
+    this.sinkEnded = false
+    this.sinkSunk = false
+
+    this.id = init.id
+    this.metadata = init.metadata ?? {}
+    this.stat = {
+      direction: init.direction,
+      timeline: {
+        open: Date.now()
+      }
+    }
+    this.maxDataSize = init.maxDataSize
+    this.onEnd = init.onEnd
+
+    this.source = this.streamSource = pushable<Uint8ArrayList>({
+      onEnd: () => {
+        // already sent a reset message
+        if (this.stat.timeline.reset !== null) {
+          this.sendCloseRead()
+        }
+
+        this.onSourceEnd()
+      }
+    })
+
+    // necessary because the libp2p upgrader wraps the sink function
+    this.sink = this.sink.bind(this)
+  }
+
+  protected onSourceEnd (err?: Error): void {
+    if (this.sourceEnded) {
+      return
+    }
+
+    this.stat.timeline.closeRead = Date.now()
+    this.sourceEnded = true
+    log.trace('%s stream %s source end - err: %o', this.stat.direction, this.id, err)
+
+    if (err != null && this.endErr == null) {
+      this.endErr = err
+    }
+
+    if (this.sinkEnded) {
+      this.stat.timeline.close = Date.now()
+
+      if (this.onEnd != null) {
+        this.onEnd(this.endErr)
+      }
+    }
+  }
+
+  protected onSinkEnd (err?: Error): void {
+    if (this.sinkEnded) {
+      return
+    }
+
+    this.stat.timeline.closeWrite = Date.now()
+    this.sinkEnded = true
+    log.trace('%s stream %s sink end - err: %o', this.stat.direction, this.id, err)
+
+    if (err != null && this.endErr == null) {
+      this.endErr = err
+    }
+
+    if (this.sourceEnded) {
+      this.stat.timeline.close = Date.now()
```

Comment:
should this be `this.stat.timeline.closeRead` ?

Reply:
This is in the `onSinkEnd` function, so we set `.close` if both the sink and the source have been closed.